### ISSUE TITLE
Feature Request: Rule option for loops and block-scoped variables

### ISSUE THREAD

**Issue by dallonf** [eslint/eslint#2517]
The [no-loop-func](http://eslint.org/docs/rules/no-loop-func) rule prevents potential issues with JavaScript's lack of block scope.

For the sake of completeness, here's one of the classic examples of this error:
```
for (var i = 0; i < 10; i++) {
  setTimeout(function() {
    console.log(i);
  }, i*100);
}
```
(this will print 10 repeatedly)

However, ES6 now has block-scoped variables, which mitigate this problem:
```
for (let i = 0; i < 10; i++) {
  setTimeout(function() {
    console.log(i);
  }, i*100);
}
```
(does the intuitive thing, printing a sequence of numbers)

But the previous syntax could still be problematic, so it's a good idea to have a rule that warns against using it. As of ESLint 0.20.0, there is no rule that would consider the former snippet a warning and allow the latter snippet.

Thoughts?

**Comment by ilyavolodin** [comment]
I'm not really sure that even in ES6 the above example is a good idea. Function creation is costly. Creating a new scope, even in block scoped environment is still a big performance hit. So I would probably still try to avoid creating functions inside the loop.

**Comment by nzakas** [comment]
I'd be up for adding an option to the rule to allow it. The purpose of the rule was to save you from the dynamically bound value of `i`, but `let` solves this problem. 

Only question is if the default behavior should be to warn on `let` loops or not.

**Comment by xjamundx** [comment]
> Only question is if the default behavior should be to warn on let loops or not.

The `no-loop-func` shouldn't warn if you're using `let`. As @nzakas pointed out the point of the rule was to prevent the async loop issue, which `let` does fix, so let's not enforce it.

I can take this on. Not even sure an option is needed.

**Comment by dallonf** [comment]
Currently, `no-loop-func` seems to check if you're making functions in a loop at all, regardless of any variables you use inside of that function. So it might not be quite a simple as it sounds...

**Comment by xjamundx** [comment]
gotcha, yeah that will be difficult. personally, I use `no-var`, so it's less of a problem :)

**Comment by nzakas** [comment]
@xjamundx sounds good.

@dallonf it would be inefficient to only warn based on which variables are being used. All of our rules try to walk this kind of being useful without needing to catch every edge case.

**Comment by hax** [comment]
Another sign is calling arrow function in the loop, which should not trigger no-loop-func in my opinion.

**Comment by mysticatea** [comment]
I suggest checking `escope.through` of functions in a loop.

The `Scope.through` is an array of references that refers to variables in upper scopes. And the reference has its variable (`Reference.resolved`) and the variable has its declaration statement (`Variable.defs`). And the declaration node has its range. 
So to check whether the declaration is `let`/`const` in the loop is probably easy.

I also can take this. :)

**Comment by nzakas** [comment]
Sounds good :)

**Comment by mysticatea** [comment]
I will try to make this.

**Comment by za-creature** [comment]
This is still broken for me as of 1.3.1:

```es6

// Works:

for (let k of [1, 2, 3, 4]) {
    (function () {
        return k;
    })();
}

// Fails:

for (let i of [1, 2, 3, 4]) {
    for (let j of [1, 2, 3, 4]) {
        (function () {
            return i + j;
        })();
    }
}
```

Seems that the current implementation only checks one level. Additionally, globals also trigger the warning ( `console.log(k)` instead of `return k` would break the first example too). Also tested on http://eslint.org/demo/

**Comment by mysticatea** [comment]
@za-creature Thank you for the report. I confirmed.

Hmm, I thought that is not related with nesting.
I thought "This rule should allow variables which are not modified."

```js
// valid

for (let i of [1, 2, 3, 4]) {
    for (let j of [1, 2, 3, 4]) {
        setTimeout(function () {
            console.log(i, j);
        }, 0);
    }
}

// Also valid

let i = 1;
for (let j of [1, 2, 3, 4]) {
    setTimeout(function () {
        console.log(i, j);
    }, 0);
}

// Invalid

for (let i of [1, 2, 3, 4]) {
    for (let j of [1, 2, 3, 4]) {
        i *= 2;
        setTimeout(function () {
            console.log(i, j);
        }, 0);
    }
}

// Also invalid

let i = 1;
for (let j of [1, 2, 3, 4]) {
    i *= 2;
    setTimeout(function () {
        console.log(i, j);
    }, 0);
}
```

This is a different enhancement from this issue. So could you open new issue?

**Comment by mrmoree** [comment]
Vote for adding an option to accept var declarations in for loops because doing so makes the code easier to read and review.
